throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,184 @@
+package throttle
+
+import (
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/loomnetwork/go-loom"
+	"github.com/loomnetwork/go-loom/plugin/contractpb"
+	"github.com/loomnetwork/loomchain"
+	"github.com/loomnetwork/loomchain/auth"
+	udw "github.com/loomnetwork/loomchain/builtin/plugins/user_deployer_whitelist"
+	"github.com/loomnetwork/loomchain/vm"
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrTxLimitReached         = errors.New("tx limit reached, try again later")
+	ErrContractNotWhitelisted = errors.New("contract not whitelisted")
+)
+
+type ContractTxLimiterConfig struct {
+	// Enables the middleware
+	Enabled bool
+	// Number of seconds each refresh lasts
+	ContractDataRefreshInterval int64
+	TierDataRefreshInterval     int64
+}
+
+func DefaultContractTxLimiterConfig() *ContractTxLimiterConfig {
+	return &ContractTxLimiterConfig{
+		Enabled:                     false,
+		ContractDataRefreshInterval: 15 * 60,
+		TierDataRefreshInterval:     15 * 60,
+	}
+}
+
+// Clone returns a deep clone of the config.
+func (c *ContractTxLimiterConfig) Clone() *ContractTxLimiterConfig {
+	if c == nil {
+		return nil
+	}
+	clone := *c
+	return &clone
+}
+
+type contractTxLimiter struct {
+	// contract_address to limiting parametres structure
+	contractToTierMap       map[string]udw.TierID
+	contractDataLastUpdated int64
+	// track of no. of txns in previous blocks per contract
+	contractToBlockTrx  map[string]*blockTxn
+	tierMap             map[udw.TierID]udw.Tier
+	tierDataLastUpdated int64
+}
+
+type blockTxn struct {
+	txn         int64
+	blockHeight int64
+}
+
+var TxLimiter *contractTxLimiter
+
+func (txl *contractTxLimiter) isAccountLimitReached(contractAddr loom.Address, curBlockHeight int64) bool {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		return false
+	}
+	// if execution reaches here => tierID and tier are valid
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID]
+	// reset contractToBlockTrx if curBlock is out of range
+	if blockTx.blockHeight <= curBlockHeight-int64(tier.BlockRange) {
```

Comment:
@Parth we have to reset state while isAccountLimitReached, a case: when after ratelimiting txns, a valid out-of-blockrange transaction comes,

about blockheight change, better we change the name from blockheight to blockRangeLowerBound, because changing the logic can create other edge cases

Reply:
@ParthDesai **